tls/s2n_psk.c

```diff
@@ -278,56 +333,104 @@ S2N_RESULT s2n_finish_psk_extension(struct s2n_connection *conn)
+
+        if (existing_psk->type == S2N_PSK_TYPE_EXTERNAL) {
+            continue;
         }
+        GUARD_AS_POSIX(s2n_array_insert_and_copy(&new_params.psk_list, new_params.psk_list.len, existing_psk));
     }
 
-    for (size_t i = 0; i < psk_vec_length; i++) {
+    if (psk_list_len != 0) {
+        notnull_check(psk_list);
+    }
+
+    for (uint32_t i = 0; i < psk_list_len; i++) {
+        notnull_check(psk_list[i]);
+
         /* Check for duplicate identities */
-        size_t array_len = conn->psk_params.psk_list.len;
-        for (size_t j = 0; j < array_len; j++) {
-            struct s2n_psk *psk = NULL;
-            GUARD_AS_POSIX(s2n_array_get(&conn->psk_params.psk_list, j, (void**) &psk));
-            ENSURE_POSIX_REF(psk);
-            if (psk->identity.size == psk_vec[i].identity_length) {
-                ENSURE_POSIX(memcmp(psk->identity.data, psk_vec[i].identity, psk->identity.size) != 0, S2N_ERR_DUPLICATE_PSK_IDENTITIES);
-            }
+        uint32_t array_len = new_params.psk_list.len;
+        for (uint32_t j = 0; j < array_len; j++) {
+            struct s2n_psk *existing_psk = NULL;
+            GUARD_AS_POSIX(s2n_array_get(&new_params.psk_list, j, (void**) &existing_psk));
+            notnull_check(existing_psk);
+
+            bool duplicate = existing_psk->identity.size == psk_list[i]->identity.size
+                    && memcmp(existing_psk->identity.data, psk_list[i]->identity.data, existing_psk->identity.size) == 0;
+            ENSURE_POSIX(!duplicate, S2N_ERR_DUPLICATE_PSK_IDENTITIES);
         }
 
         struct s2n_psk *new_psk = NULL;
-        GUARD_AS_POSIX(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &new_psk));
-        ENSURE_POSIX_REF(new_psk);
-        GUARD(s2n_psk_init(new_psk, S2N_PSK_TYPE_EXTERNAL));
-        GUARD(s2n_psk_new_identity(new_psk, psk_vec[i].identity, psk_vec[i].identity_length));
-        GUARD(s2n_psk_new_secret(new_psk, psk_vec[i].secret, psk_vec[i].secret_length));
-        GUARD_AS_POSIX(s2n_psk_set_hmac(new_psk, psk_vec[i].hmac));
+        GUARD_AS_POSIX(s2n_array_pushback(&new_params.psk_list, (void**) &new_psk));
+        notnull_check(new_psk);
+
+        GUARD_AS_POSIX(s2n_psk_init(new_psk, S2N_PSK_TYPE_EXTERNAL));
+        ENSURE_POSIX(s2n_psk_set_identity(new_psk, psk_list[i]->identity.data, psk_list[i]->identity.size) == S2N_SUCCESS,
+                S2N_ERR_INVALID_ARGUMENT);
+        ENSURE_POSIX(s2n_psk_set_secret(new_psk, psk_list[i]->secret.data, psk_list[i]->secret.size) == S2N_SUCCESS,
+                S2N_ERR_INVALID_ARGUMENT);
+        new_psk->hmac_alg = psk_list[i]->hmac_alg;
     }
 
+    /* Verify PSK list will fit in the ClientHello pre_shared_key extension */
+    if (conn->mode == S2N_CLIENT) {
+        uint32_t offered_psks_size = 0;
+        GUARD_AS_POSIX(s2n_psk_parameters_offered_psks_size(&new_params, &offered_psks_size));
+        ENSURE_POSIX(offered_psks_size <= S2N_MAX_OFFERED_PSK_LIST_SIZE, S2N_ERR_PSKS_TOO_LONG);
```

Comment:
The use case for the error code does not clearly match its description, this is rather confusing to me.  Maybe update the error code name to be more specific?

Reply:
We're comparing the size of the "offered psks" to their maximum size. In what way does that not match the error description string? What description string would you suggest?